gh-80642: timeit - make target time of autorange configurable (#140283)
python/cpython: Doc/library/timeit.rst

The module defines three convenience functions and a public class:

.. function:: timeit(stmt='pass', setup='pass', timer=<default timer>, number=1000000, globals=None, target_time=0.2)

   Create a :class:`Timer` instance with the given statement, *setup* code and
   *timer* function and run its :meth:`.timeit` method with *number* executions.
   The optional *globals* argument specifies a namespace in which to execute the
   code. If *number* is 0, :meth:`.autorange` method is executed, a convenience

>    function that calls :meth:`.timeit` repeatedly so that the total time >= *target_time* second.

Reviewer: Please wrap to 79 characters.

Author: Done

   .. versionchanged:: 3.5
      The optional *globals* parameter was added.

>    .. versionchanged:: 3.8

Reviewer 2: By now, this would go into 3.15. We can use `next` and it'll be replaced with whatever version is eventually gets merged into:

```suggestion
   .. versionchanged:: next
```

Reviewer: Ah, we duplicated these suggestions.

>    .. versionchanged:: 3.8

Reviewer: ```suggestion
   .. versionchanged:: next
```

      The optional *target_time* parameter was added.

.. function:: repeat(stmt='pass', setup='pass', timer=<default timer>, repeat=5, number=1000000, globals=None, target_time=0.2)

   Create a :class:`Timer` instance with the given statement, *setup* code and
   *timer* function and run its :meth:`.repeat` method with the given *repeat*
   count and *number* executions.  The optional *globals* argument specifies a
   namespace in which to execute the code. If *number* is 0, :meth:`.autorange`
   method is executed, a convenience function that calls :meth:`.timeit` repeatedly
   so that the total time >= *target_time* second.

   .. versionchanged:: 3.5
      The optional *globals* parameter was added.

   .. versionchanged:: 3.7
      Default value of *repeat* changed from 3 to 5.

>    .. versionchanged:: 3.15

Reviewer: ```suggestion
   .. versionchanged:: next
```

>    .. versionchanged:: 3.15

Reviewer 2: ```suggestion
   .. versionchanged:: next
```

   Create a :class:`Timer` instance with the given statement, *setup* code and
   *timer* function and run its :meth:`.timeit` method with *number* executions.
   The optional *globals* argument specifies a namespace in which to execute the
   code. If *number* is 0, :meth:`.autorange` method is executed, a convenience
   function that calls :meth:`.timeit` repeatedly so that the total time >=
   *target_time* second.

   .. versionchanged:: 3.5
      The optional *globals* parameter was added.

   .. versionchanged:: next
      The optional *target_time* parameter was added.

.. function:: repeat(stmt='pass', setup='pass', timer=<default timer>, repeat=5, number=1000000, globals=None, target_time=0.2)

   Create a :class:`Timer` instance with the given statement, *setup* code and
   *timer* function and run its :meth:`.repeat` method with the given *repeat*
   count and *number* executions.  The optional *globals* argument specifies a
   namespace in which to execute the code. If *number* is 0, :meth:`.autorange`
   method is executed, a convenience function that calls :meth:`.timeit`

>    repeatedly so that the total time >= *target_time* second.

Reviewer: ```suggestion
   namespace in which to execute the code. If *number* is 0, the :meth:`autorange`
   method is executed, and a convenience function calls :meth:`timeit`
   repeatedly so that the total time >= *target_time* seconds.
```
The current wording is quite odd IMO.

   .. versionchanged:: 3.5
      The optional *globals* parameter was added.

>    .. versionchanged:: 3.8

Reviewer 2: ```suggestion
   .. versionchanged:: next
```

            timeit.Timer('for i in range(10): oct(i)', 'gc.enable()').timeit()

   .. method:: Timer.autorange(callback=None, target_time=None)

      Automatically determine how many times to call :meth:`.timeit`.

      This is a convenience function that calls :meth:`.timeit` repeatedly
      so that the total time >= *Timer.target_time* seconds, returning the eventual
      (number of loops, time taken for that number of loops). It calls
      :meth:`.timeit` with increasing numbers from the sequence 1, 2, 5,
      10, 20, 50, ... until the time taken is at least *target_time* seconds.

      If *callback* is given and is not ``None``, it will be called after
      each trial with two arguments: ``callback(number, time_taken)``.

      .. versionadded:: 3.6

      .. versionchanged:: next
         The optional *target_time* parameter was added.

> 

Reviewer: ```suggestion

```
Please preserve the spacing

   .. versionadded:: 3.5

> .. cmdoption:: -t, --target_time=T

Reviewer 2: ```suggestion
.. option:: -t, --target-time=T
```

    calls :meth:`.timeit` repeatedly so that the total time >= *target_time* seconds

>    .. versionadded:: 3.15

Reviewer: ```suggestion
   .. versionadded:: next
```

>    .. versionadded:: 3.15

Reviewer 2: ```suggestion
   .. versionadded:: next
```

> .. cmdoption:: -v, --verbose

Reviewer: Why was this changed to `.. cmdoption::`? Please remove unrelated changes, there is a separate issue for this.

Reviewer 2: I don't think this needs a separate issue, but the [Sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-references) says `cmdoption` is deprecated so perhaps we should keep `option` here and also use it for the new one above?

Reviewer: I meant that there *already is* an issue on the topic: https://github.com/python/cpython/issues/130160

I agree with keeping and using `option`.

Author: It was a mistake I made while merging main to the original PR. Changed it back to `option`.
